debian/rules

```diff
@@ -0,0 +1,102 @@
+	[ -d vendor/ ] || go mod vendor
+
+	# Some crates are shipped with .a files, which get removed by the helpers during the package build as a safety measure.
+	# This results in cargo failing to compile, since the files (which are listed in the checksums) are not there anymore.
+	# For those crates, we need to replace their checksum with a more general one that only lists the crate checksum, instead of each file.
+	if [ ! -d $(CARGO_VENDOR_DIR)/ ]; then \
+		$(CARGO) vendor $(CARGO_VENDOR_DIR); \
+		[ ! -e $(DH_CARGO_VENDORED_SOURCES) ] || $(DH_CARGO_VENDORED_SOURCES); \
+		\
+		[ -e /usr/bin/jq ] || (echo "jq is required to run this script. Try installing it with 'sudo apt install jq'" && exit 1); \
+		\
+		for dep in $$(ls vendor_rust -1); do \
+			checksum_file="vendor_rust/$${dep}/.cargo-checksum.json"; \
+			a_files=$$(cat $${checksum_file} | jq '.files | keys | map(select(.|test(".a$$")))'); \
+			if [ "$$a_files" = "[]" ]; then \
+				continue; \
+			fi; \
+			pkg_checksum=$$(cat "$${checksum_file}" | jq '.package'); \
+			echo "{\"files\": {}, \"package\": $${pkg_checksum}}" >"$${checksum_file}"; \
+		done; \
+	fi
+
+override_dh_auto_configure:
+	dh_auto_configure
+
+	[ ! -e $(DH_CARGO_VENDORED_SOURCES) ] || $(DH_CARGO_VENDORED_SOURCES)
+	DEB_HOST_GNU_TYPE=$(DEB_HOST_GNU_TYPE) DEB_HOST_RUST_TYPE=$(DEB_HOST_RUST_TYPE) \
+		CARGO_HOME=$(CURDIR)/debian/cargo_home DEB_CARGO_CRATE=nss_$(shell dpkg-parsechangelog --show-field Version) \
+		RUSTFLAGS="--cfg=rustix_use_libc" \
+		$(CARGO) prepare-debian $(CARGO_VENDOR_DIR)
+
+override_dh_auto_build:
+	# Build PAM library & Go loader
+	go generate -x ./pam
+
+	# Build the NSS library
+	DEB_HOST_GNU_TYPE=$(DEB_HOST_GNU_TYPE) \
+	DEB_HOST_RUST_TYPE=$(DEB_HOST_RUST_TYPE) \
+	CARGO_HOME=$(CURDIR)/debian/cargo_home \
+	$(CARGO) build --release
+
+	# Build the CLI
+	DH_GOLANG_BUILDPKG=github.com/ubuntu/authd/cmd/authd dh_auto_build
+
+override_dh_auto_test:
+	# We need to specify these Rust related variables to the Go tests in order to build the NSS lib
+	# with the cargo wrapper in the integration tests in order to force cargo to use vendored deps
+	# instead of querying crates.io for them.
+	DEB_HOST_GNU_TYPE=$(DEB_HOST_GNU_TYPE) \
+	DEB_HOST_RUST_TYPE=$(DEB_HOST_RUST_TYPE) \
+	CARGO_HOME=$(CURDIR)/debian/cargo_home \
+	CARGO_PATH=$(CARGO) \
+	dh_auto_test
+
+override_dh_auto_install:
+	dh_auto_install -- --no-source
+
+	# Install daemon in /sbin
+	mv debian/authd/usr/bin/ debian/authd/usr/sbin
+
+	# Install PAM module configuration
+	mkdir -p debian/authd/usr/share/pam-configs
+	cp pam/go-loader/libpam-authd.pam-auth-update debian/authd/usr/share/pam-configs/authd
+
+	# Install PAM
+	mkdir -p debian/authd/usr/lib/$(DEB_HOST_GNU_TYPE)/security
+	cp pam/go-loader/pam_go_loader.so pam/pam_authd.so debian/authd/usr/lib/$(DEB_HOST_GNU_TYPE)/security/
+
+	# Install NSS
+	cp target/$(DEB_HOST_RUST_TYPE)/release/libnss_authd.so debian/authd/usr/lib/$(DEB_HOST_GNU_TYPE)/libnss_authd.so.2
```

Comment:
Is there a reason why we are using `debian/authd` instead of `debian/tmp` here? I'm definitely not an ace in packaging, so it's a very honest question (I'm not even sure it matters).

Also, I noticed you are using `cp` to copy the files. Wouldn't `install` be better? This way we can also fix the file permissions during the copying.

Reply:
> The advantage of using `debian/authd` over `debian/tmp` is that this removes the need for the `install` file, since we already know what files belong to the package. As we have a single package I feel this is simpler 🤔.

Yeah sure, but having a `debian/install` file also ensures that everything that we generated got installed.

Now, that's not a problem per se now since we've just one package, but it would be better in case of splitting, and in general it makes things easier if your package name gets renamed.

However, speaking if it... I'm not sure if we want that, but generally the pam libraries are packaged as different package. I'm not sure if that's due to some rule or good behavior, but I'm wondering if we should do this here too.